src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -139,35 +140,55 @@ internal PEReader PEReaderOpt
             }
         }
 
-        internal unsafe MetadataReader MetadataReader
+        internal MetadataReader MetadataReader
         {
             get
             {
                 if (_lazyMetadataReader == null)
                 {
-                    MetadataReader newReader;
+                    InitializeMetadataReader();
+                }
 
-                    // PEModule is either created with metadata memory block or a PE reader.
-                    if (_metadataPointerOpt != IntPtr.Zero)
-                    {
-                        newReader = new MetadataReader((byte*)_metadataPointerOpt, _metadataSizeOpt, MetadataReaderOptions.ApplyWindowsRuntimeProjections, StringTableDecoder.Instance);
-                    }
-                    else
-                    {
-                        Debug.Assert(_peReaderOpt != null);
-                        if (!_peReaderOpt.HasMetadata)
-                        {
-                            throw new BadImageFormatException(CodeAnalysisResources.PEImageDoesntContainManagedMetadata);
-                        }
+                if (_isDisposed)
+                {
+                    // Without locking, which might be expensive, we can't guarantee that the underlying memory 
+                    // won't be accessed after the metadata object is disposed. However we can do a cheap check here that 
+                    // handles most cases.
+                    ThrowMetadataDisposed();
+                }
 
-                        newReader = _peReaderOpt.GetMetadataReader(MetadataReaderOptions.ApplyWindowsRuntimeProjections, StringTableDecoder.Instance);
-                    }
+                return _lazyMetadataReader;
+            }
+        }
 
-                    Interlocked.CompareExchange(ref _lazyMetadataReader, newReader, null);
+        private unsafe void InitializeMetadataReader()
+        {
+            MetadataReader newReader;
+
+            // PEModule is either created with metadata memory block or a PE reader.
+            if (_metadataPointerOpt != IntPtr.Zero)
+            {
+                newReader = new MetadataReader((byte*)_metadataPointerOpt, _metadataSizeOpt, MetadataReaderOptions.ApplyWindowsRuntimeProjections, StringTableDecoder.Instance);
+            }
+            else
+            {
+                Debug.Assert(_peReaderOpt != null);
+
+                // Checking for an empty image is a workaround for https://github.com/dotnet/corefx/issues/1815: 
+                if (_peReaderOpt.IsEntireImageAvailable && _peReaderOpt.GetEntireImage().Length == 0 || !_peReaderOpt.HasMetadata)
```

Comment:
GetEntireImage causes the entire image to be memory mapped. This causes a VM regression in the Picasso "Start Debugging scenario."